Introduce new command line runner for Twisted: twist

wsanchez opened the pull request:
Introduce new command line runner for Twisted: `twist`: https://twistedmatrix.com/trac/ticket/5705.

```
Usage: twist [options] plugin [plugin_options]
Options:
      --log-level=   Set default log level. (options: "debug", "info", "warn",
                     "error", "critical"; default: "info")
      --help         Display this help and exit.
      --version      Print version and exit.
      --log-format=  Log file format. (options: "text", "json"; default: "text"
                     if the log file is a tty, otherwise "json")
      --log-file=    Log to file. ("-" for stdout, "+" for stderr; default: "-")
      --reactor=     The name of the reactor to use. (options: "win32", "gi",
                     "kqueue", "poll", "iocp", "select", "epoll", "gtk2",
                     "gtk3", "wx", "default", "cf", "glib2")

Run a Twisted application.
Commands:
    conch            A Conch SSH service.
    dns              A domain name server.
    ftp              An FTP server.
    inetd            An inetd(8) replacement.
    mail             An email service
    manhole          An interactive remote debugger service accessible via
                     telnet and ssh and providing syntax coloring and basic line
                     editing functionality.
    news             A news server.
    portforward      A simple port-forwarder.
    procmon          A process watchdog / supervisor
    socks            A SOCKSv4 proxy service.
    web              A general-purpose web server which can serve from a
                     filesystem or application resource.
    words            A modern words server
    xmpp-router      An XMPP Router server
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/twisted/twisted/134)

<!-- Reviewable:end -->

wsanchez committed:
Add exit() and ExitStatus constants.

wsanchez committed:
Add empty twisted.application.runner module.

wsanchez committed:
It's a stub, not a mock.  Whatevs, but fix.

wsanchez committed:
Add Runner, RunnerOptions

wsanchez committed:
Add DummyExit for fancy

wsanchez committed:
Patching sys.exit feels sketchy, so don't.

wsanchez committed:
Implement IReactorCore in MemoryReactor

wsanchez committed:
Don't think these make sense.

wsanchez committed:
Document RunnerOptions.
logFileFormat isn't used.

wsanchez committed:
Test whenRunning and reactorExited.
Test that whenRunning is called when the reactor starts.

wsanchez committed:
Remove commented out imports

wsanchez committed:
Test installing and running the default reactor.

wsanchez committed:
default is the default reactor

wsanchez committed:
Test starting specified reactor.

wsanchez committed:
execute -> run

wsanchez committed:
doc

wsanchez committed:
Let exit do the writing

wsanchez committed:
Add --version

wsanchez committed:
epytext cleanup

wsanchez committed:
More quotes

wsanchez committed:
Add --pidfile

wsanchez committed:
Add --log-level.
Toss out --list-reactors.
Add options and defaults to doc strings.

wsanchez committed:
Implement Twistd.
It has an option to list available reactors, an option to select a reactor.
It runs and stops the reactor.

wsanchez committed:
Don't set logFile in hack code

wsanchez committed:
parseArgs -> postOptions

wsanchez committed:
Add --log-file

wsanchez committed:
Don't override user-specified FLO

wsanchez committed:
Add --log-format

wsanchez committed:
Change exit code from EX_USAGE to EX_CONFIG

wsanchez committed:
Choose default log format based on whether log file is a tty.

wsanchez committed:
run subcommands

wsanchez committed:
Factor into bits

wsanchez committed:
Doc strings

wsanchez committed:
Make open overridable

wsanchez committed:
Put loadedPlugins into dict, not attr

wsanchez committed:
Raise UsageError instead of exiting directly when invalid arguments are given

wsanchez committed:
Add twist script.

wsanchez committed:
Call startService, not privilegedStartService.

wsanchez committed:
Add plugins property

wsanchez committed:
unused

wsanchez committed:
cosmetic

wsanchez committed:
Add tests for TwistOptions

wsanchez committed:
Keep track of added system event triggers

wsanchez committed:
More tests

wsanchez committed:
Make stdout pretend to be a tty

wsanchez committed:
Test startService

wsanchez committed:
Ask for json so it's always doing the same thing regardless of tty or not

wsanchez committed:
patch MultiService.startService

wsanchez committed:
Add test_main

wsanchez committed:
unused

wsanchez committed:
cosmetic

wsanchez committed:
docstrings

wsanchez committed:
lint

wsanchez committed:
implementer, not implements

wsanchez committed:
lint docstrings

wsanchez committed:
Fix epytext errors

wsanchez committed:
Implement test_startLogging

wsanchez committed:
Subclass from FilePath so that twistedchecker stops complaining

wsanchez committed:
Document exit codes.

rodrigc commented:
@wsanchez Please add a topfile as specified in: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

 so that https://travis-ci.org/twisted/twisted/builds/135205034 does not break

wsanchez committed:
Add test for unknown reactor raising NoSuchReactor, which should get catched and UsageError raised. (+1 squashed commit)
Squashed commits:
[08f96b9] Add test for unknown reactor raising NoSuchReactor, which should get catched and UsageError raised.

wsanchez committed:
Add test for unable to open log file case due to EnvironmentError.
Use EX_IOERR instead of EX_CANTCREAT since the error might be be able creation.

wsanchez committed:
pass is unneeded here

wsanchez committed:
No need for encoding games here; it's a number.

wsanchez committed:
Add test for handling of bogus PID file.

wsanchez committed:
Add test for handle of case where PID file is not readable.
Use IOERR for not readable case.
Capitalize PID.

wsanchez committed:
Add test for empty PID file

wsanchez committed:
Remove __main__ thingo

wsanchez committed:
Tests for checking that messages are written to stdout or stderr based on status code.

wsanchez committed:
Be more specific about the options mapping

wsanchez committed:
Make runner API private for now.

wsanchez committed:
Revert change

wsanchez committed:
Add news file
